test/e2e/network/traffic_policy.go

```diff
@@ -0,0 +1,469 @@
+	} else {
+		internalClientNodeName = tr.endpointNode.Name
+		externalClientNodeName = peerNode.Name
+	}
+	// create client pods
+	podClient := e2epod.NewPodClient(f)
+	tr.internalClientPod = createClientPod(podClient, tr.clusterIPEndpoint, f.Namespace.Name, internalClientNodeName, "internal")
+	tr.externalClientPod = createClientPod(podClient, tr.nodeportEndpoint, f.Namespace.Name, externalClientNodeName, "external")
+
+	return &tr
+}
+
+// seenPodsCheckFunc is a function to be used by checkEndpointHost below.
+// checkEndpointHost runs several curls to an endpoint and calls this function to validate the pods which handled the requests.
+// seenPodsCheckFunc need not check an expected zone because it is part of checkEndpointHost to verify expectedZone (if specified) after each request
+// seenPodsCheckFunc should return true if all the handling pods are expected, and false if an invalid result is found.
+type seenPodsCheckFunc func(tr *trafficPolicyTestResources, seenPods map[string]string) bool
+
+// This is used with 'Cluster' traffic policies within a zone (topology hints: auto)
+func sawMoreThanOnePod(_tr *trafficPolicyTestResources, seenPods map[string]string) bool {
+	return len(seenPods) > 1
+}
+
+// This is used when 'Local' traffic policy is specified to verify the expected node
+func sawOnePodOnEndpointNode(tr *trafficPolicyTestResources, seenPods map[string]string) bool {
+	if len(seenPods) != 1 {
+		return false
+	}
+	for pod := range seenPods {
+		node, ok := tr.nodesByPod[pod]
+		return ok && node == tr.endpointNode.Name
+	}
+	return false
+}
+
+// This is used with 'Cluster' traffic policies with topology hints disabled
+func sawMultipleZones(_tr *trafficPolicyTestResources, seenPods map[string]string) bool {
+	zones := make(map[string]bool)
+	for _, zone := range seenPods {
+		zones[zone] = true
+	}
+	return len(seenPods) > 1 && len(zones) > 1
+}
+
+func createClientPod(podClient *e2epod.PodClient, endpoint string, namespaceName string, nodeName string, podSuffix string) *v1.Pod {
+	ginkgo.By("creating a client pod for probing the service from " + nodeName)
+	// put a pod in the single node zone to use for curling
+	podName := "curl-from-" + nodeName + podSuffix
+
+	/// turn this into a function which takes endpoint, and expected reply details
+	clientPod := e2epod.NewAgnhostPod(namespaceName, podName, nil, nil, nil, "serve-hostname")
+	nodeSelection := e2epod.NodeSelection{Name: nodeName}
+	e2epod.SetNodeSelection(&clientPod.Spec, nodeSelection)
+
+	ginkgo.By("Checking access to endpoint " + endpoint)
+
+	cmd := fmt.Sprintf(`date; for i in $(seq 1 3000); do sleep 0.5; echo "Date: $(date) Try: ${i}"; curl -q -s --connect-timeout 2 http://%s/hostname ; echo; done`, endpoint)
+	clientPod.Spec.Containers[0].Command = []string{"/bin/sh", "-c", cmd}
+	clientPod.Spec.Containers[0].Name = clientPod.Name
+	return podClient.CreateSync(clientPod)
+}
+
+// checkEndpointHost creates a pod on the specified node.
```

Comment:
this comment seems outdated, the client is passed as an argument now.

Maybe you can convert this to two methods of `trafficPolicyTestResources` , as `checkEndpointHostInternal` and `checkEndpointHostExternal`, sharing the common code and using the corresponding pod in each case

Reply:
Good call. I pushed an update with this change.